components/event-bus/cmd/event-publish-service/publisher/publisher.go

```diff
@@ -35,62 +42,81 @@ func NewKnativePublisher() KnativePublisher {
-	string) {
+	string, string) {
+
 	// knativelib should not be nil
 	if knativeLib == nil {
 		log.Println("knative-lib is nil")
-		return api.ErrorResponseInternalServer(), FAILED
-	}
-
-	// channelName should be present
-	if channelName == nil || len(*channelName) == 0 {
-		log.Println("channelName is missing")
-		return api.ErrorResponseInternalServer(), FAILED
+		return api.ErrorResponseInternalServer(), FAILED, empty
 	}
 
 	// namespace should be present
 	if namespace == nil || len(*namespace) == 0 {
 		log.Println("namespace is missing")
-		return api.ErrorResponseInternalServer(), FAILED
+		return api.ErrorResponseInternalServer(), FAILED, empty
 	}
 
 	// headers should be present
 	if headers == nil || len(*headers) == 0 {
 		log.Println("headers are missing")
-		return api.ErrorResponseInternalServer(), FAILED
+		return api.ErrorResponseInternalServer(), FAILED, empty
 	}
 
 	// payload should be present
 	if payload == nil || len(*payload) == 0 {
 		log.Println("payload is missing")
-		return api.ErrorResponseInternalServer(), FAILED
+		return api.ErrorResponseInternalServer(), FAILED, empty
+	}
+
+	if len(source) == 0 || len(eventType) == 0 || len(eventTypeVersion) == 0 {
+		log.Println("one of the event source, type or version value is missing")
+		return api.ErrorResponseInternalServer(), FAILED, empty
 	}
 
-	// get the knative channel
-	channel, err := knativeLib.GetChannel(*channelName, *namespace)
+	//Adding the event-metadata as channel labels
+	knativeChannelLabels := make(map[string]string)
+	knativeChannelLabels[subscriptionSourceID] = source
+	knativeChannelLabels[subscriptionEventType] = eventType
+	knativeChannelLabels[subscriptionEventTypeVersion] = eventTypeVersion
+
+	//Fetch knative channel via label
+	channel, err := knativeLib.GetChannelByLabels(*namespace, &knativeChannelLabels)
 	if err != nil {
-		log.Printf("cannot find the knative channel '%v' in namespace '%v'", *channelName, *namespace)
+		log.Printf("an error occured while trying to get the knative channel for source: '%v', event-type: '%v', event-type-version: '%v' in namespace '%v'\n"+
+			"error: '%v':", source, eventType, eventTypeVersion, *namespace, err)
 		log.Println("incrementing ignored messages counter")
 		metrics.TotalPublishedMessages.With(prometheus.Labels{
 			metrics.Namespace:        *namespace,
 			metrics.Status:           IGNORED,
 			metrics.SourceID:         source,
 			metrics.EventType:        eventType,
 			metrics.EventTypeVersion: eventTypeVersion}).Inc()
-		return nil, IGNORED
+		return nil, IGNORED, empty
+	}
+
+	return publisher.publishOnChannel(knativeLib, channel, namespace, headers, payload)
+}
+
+func (publisher *DefaultKnativePublisher) publishOnChannel(knativeLib *knative.KnativeLib, channel *messagingV1Alpha1.Channel, namespace *string, headers *map[string][]string,
+	payload *[]byte) (*api.Error, string, string) {
+
+	// knative Channel reference should not be nil
+	if channel == nil {
```

Comment:
channel will never be `nil` as in GetChannelByLabels() error is thrown if no channels are found